Provide capability to modify the meta field type

hickeyma opened the pull request:
Currently in Kibana you are unable to change the field type of a meta field. This PR is to update the meta field to be a field type where you create and update the field similar to a scripted field. This gives you the option to modify the field from the default string type.

Fixes #5923

elasticmachine commented:
Can one of the admins verify this patch?

epixa commented:
@hickeyma Can you update the description of this PR with details about the change?  That will help reviewers understand the change and could help us describe the issue in a release blog post.

hickeyma commented:
@epixa Sure, updated now.

spalger commented:
A couple notes:

 1. Looks like refreshing the fields causes the default meta-fields to be duplicated

 1. I think there needs to be a way to delete a meta field (including the defaults) after creating it.
     - Refreshing the field list shouldn't recreate the default meta fields

 1. meta-fields for existing index patterns should be shown in the meta-field list even if they were created before this change

 1. meta-fields should probably be filters out of the standard-field list

hickeyma commented:
Thanks for feedback @spalger. I will refactor the comments when I get a chance.

hickeyma commented:
Thanks for the [notes](https://github.com/elastic/kibana/pull/8554#issuecomment-256744437) @spalger. Sorry for delay in replying but was OOTO last week.

I have some feedback on your notes below.

>Looks like refreshing the fields causes the default meta-fields to be duplicated

I have a fix for this but I think this issue is better solved by your next point.
>I think there needs to be a way to delete a meta field (including the defaults) after creating it

So, when once an old style meta field is now loaded as the new meta field type then the old meta field should be deleted from elasticsearch. Is this what your are thinking?

>meta-fields for existing index patterns should be shown in the meta-field list even if they were created before this change

Is that not already working?

>meta-fields should probably be filters out of the standard-field list 

Not sure what you mean by this. Could you elaborate further?

>adding a field like _uid does not check whether it's aggregatable or searchable, which means they default to false and are not usable in visualize or filtering.

Are you thinking that we add a control to the meta field page where you can enable/disable `aggregatable` or `searchable`?

hickeyma commented:
@spalger I have pushed [commit](https://github.com/elastic/kibana/pull/8554/commits/c5f89ce79ab9e520cc72c76fa4b0523b9e1042cb). It needs more work to it but I wanted to use it as basis for further discussion with you to see if goign in the right direction.

spalger commented:
I need to take a deeper look tomorrow but this is looking great @hickeyma. 

Responses to your previous questions:

> So, when once an old style meta field is now loaded as the new meta field type then the old meta field should be deleted from elasticsearch. Is this what your are thinking?

Well, if I am reading the correctly then all we need to do is add a `meta` property to older fields in order to make them "new" metafields. Is that right? No need to delete anything from elasticsearch.

What I meant by "I think there needs to be a way to delete a meta field" is that I think there should be a delete button for each metafield, just like there are for scripted fields.

![image](https://cloud.githubusercontent.com/assets/1329312/20290586/2adc5bda-aa9e-11e6-9809-c26d428b6caa.png)

> Does this mean meta fields should still be visible in the fields tab?

I don't think that metafields should ever be visible on the main fields tab. They should only live in the metafields tab.

>> meta-fields for existing index patterns should be shown in the meta-field list even if they were created before this change

> Is that not already working?

Doesn't look like it. The functionality here requires that fields have the `meta` property to be considered meta fields. Index patterns created before this pr won't have the `meta` property set.

> Are you thinking that we add a control to the meta field page where you can enable/disable aggregatable or searchable?

Well, elasticsearch does tell us about different whether we can search or aggregate on different meta fields => `GET logstash-*/_field_stats?fields=_*`

hickeyma commented:
Some follow up @spalger..

>Well, if I am reading the correctly then all we need to do is add a meta property to older fields in order to make them "new" metafields. Is that right? No need to delete anything from elasticsearch.

I think that these are settings and not fields. I therefore [deleted the settings](https://github.com/hickeyma/kibana/blob/10c5a76311e561f00daaa19984a7b32c5983c393/src/ui/public/index_patterns/_transform_mapping_into_fields.js#L53) once they become fields.

>What I meant by "I think there needs to be a way to delete a meta field" is that I think there should be a delete button for each metafield, just like there are for scripted fields.

Sure, I can add the button.

>I don't think that metafields should ever be visible on the main fields tab. They should only live in the metafields tab.

Ok, will need to stop them appearing in the main field tab.

> Doesn't look like it. The functionality here requires that fields have the meta property to be considered meta fields. Index patterns created before this pr won't have the meta property set.

At the moment, all meta fields in the index pattern, (e.g. _id, _type etc.) and any meta fields added in "Advanced Settings" will  be added as a meta field in the system. Is this what you are referring to?

>Well, elasticsearch does tell us about different whether we can search or aggregate on different meta fields => GET logstash-*/_field_stats?fields=_*

I have added a checkbox for aggregatable and searchable on the meta field page where you can set those attributes for the field.

spalger commented:
> I think that these are settings and not fields. I therefore deleted the settings once they become fields.

That's not going to work for people with more than one index pattern, right?

How do you feel about using this logic?

 1. add a `metaFieldsVersion: 'number'` field to the [index pattern object mapping](https://github.com/elastic/kibana/blob/a0a1fbda56156903abcd77186d98396b4f3e8872/src/ui/public/index_patterns/_index_pattern.js#L40-L57)
     1. this field should be set to `1` for any index pattern that has the meta fields
     1. any index pattern that doesn't have a `metaFieldsVersion` will need to be upgraded when it's loaded

 1. Remove the metaField logic from [_transform_mapping_into_fields.js](https://github.com/elastic/kibana/blob/0e7b5c7b0678e78b3ed53c0941200cb5c536fa78/src/ui/public/index_patterns/_transform_mapping_into_fields.js)

 1. When [refreshing the field list](https://github.com/elastic/kibana/blob/0e7b5c7b0678e78b3ed53c0941200cb5c536fa78/src/ui/public/index_patterns/_index_pattern.js#L362-L367)
     1. If the `metaFieldsVersion` number is 1:
         1. extract just the meta fields from the old field list
         1. add the metaFields to the new field list [like we do for scripted fields](https://github.com/elastic/kibana/blob/0e7b5c7b0678e78b3ed53c0941200cb5c536fa78/src/ui/public/index_patterns/_index_pattern.js#L152)
         1. Save
     1. If the `metaFieldsVersion` is not set:
         1. Add a new metaField to the field list for each field in `config.get('metaFields')`
         1. Set the `metaFieldsVersion` to 1
         1. Save

 1. When loading an index pattern:
     1. if the `metaFieldsVersion` is not set:
         1. For each field in `config.get('metaFields')`
            1. If the indexPattern doesn't have that field do nothing
            1. if the indexPattern *does* have a field with that name:
                1. Set the fields `meta` property to `true`
         1. Set the `metaFieldsVersion` to 1
         1. Save the indexPattern back to elasticsearch

hickeyma commented:
@spalger, I have done an update in the following [commit](https://github.com/elastic/kibana/pull/8554/commits/e71ad4c431452f86e9f842aa48afbb65c891e94d). 
I haven't tackled your feedback on [changing the logic](https://github.com/elastic/kibana/pull/8554#issuecomment-262611348) yet. I wanted to clean up a few issues first and add some missing capability in the view. I will look at changing the logic as it would seem to not be able to handle an upgrade of an existing systems with indices already created.

hickeyma commented:
@spalger, re [comment](https://github.com/elastic/kibana/pull/8554#issuecomment-262611348):

I have updated the logic. Here are some caveats:

> add a `metaFieldsVersion:` 'number' field

I have used a `boolean` instead.

>Remove the metaField logic from `_transform_mapping_into_fields.js`

Not sure that this can be removed as logic in  [index pattern](https://github.com/elastic/kibana/blob/master/src/ui/public/index_patterns/_index_pattern.js) depends on the pattern and fields to be created for updates.

>When [refreshing the field list](https://github.com/elastic/kibana/blob/master/src/ui/public/index_patterns/_index_pattern.js#L362-L367)

At the moment, I am more sided on doing it the same way regardless of `metaFieldsVersion` set or not.

> When loading an index pattern

I have gone with doing a refresh, instead of a new function to do it. Not sure adding a new function adds a great advantage.

ycombinator commented:
Hey @hickeyma, nice to "meet" you. I'm going to be playing 2nd reviewer on this PR in addition to @spalger. I'm catching up on the conversation so far and will be pulling down this branch to play with it presently. Anyway, just a heads up!

hickeyma commented:
Welcome onboard @ycombinator! :-)

ycombinator commented:
@hickeyma, I just did a review of the functionality introduced by this PR. It's looking good, with a couple of minor issues:

1. If you create a _scripted_ field, then go over to the meta fields tab and delete any field, the scripted field is what actually gets deleted (but not the meta field you just tried to delete). If you then delete another meta field, it gets deleted as expected.

2. After a user creates a new meta field, it would be nice to bring the user back to the meta fields tab (akin to what happens when a user creates a new scripted field). Same thing with updating an existing meta field.

3. I edited a meta field, unchecked the searchable and aggregatable, and clicked the Update Field button to save my changes. When I go back to the meta fields tab, those changes aren't there. I checked the corresponding document in `.kibana` and they aren't there either so for some reason they aren't getting saved. FWIW, other edits (such as changing the format of the meta field) seem to get saved properly.

hickeyma commented:
@ycombinator thanks for https://github.com/elastic/kibana/pull/8554#issuecomment-274519466 . I have updated as follows:

>1. If you create a scripted field, then go over to the meta fields tab and delete any field, the scripted field is what actually gets deleted 

Fixed

>2. After a user creates/updates a new meta field, it would be nice to bring the user back to the meta fields tab 

Fixed

>3. I edited a meta field, unchecked the searchable and aggregatable, and clicked the Update Field button to save my changes.

This was a misunderstanding on my part. These fields should not have been shown as 'check boxes' and therefore modifiable. They are fields which are set when the field is initiated depending on the field typo(i.e. meta, scripted etc.). I have removed the 'check boxes'.

ycombinator commented:
Hi @hickeyma, those changes look great! I think we're pretty close now to getting this merged. I'm reviewing the code now...

ycombinator commented on src/ui/public/index_patterns/_index_pattern.js:
```diff
@@ -110,7 +113,11 @@ export default function IndexPatternFactory(Private, Notifier, config, kbnIndex,
       promise = indexPattern.refreshFields();
+    } else if (!indexPattern.metaFieldsVersion) {
+      promise = indexPattern.refreshFields();
+      indexPattern.metaFieldsVersion = true;
```
This is related to my comment [earlier in this file](https://github.com/elastic/kibana/pull/8554/files#r98642125). If we make `indexPattern.metaFieldsVersion` a number, then it should probably be initialized to 1 over here, for this initial implementation of meta fields.

hickeyma replied:
Done.

ycombinator commented:
Found one more minor functional issue: when I delete a meta field from the meta fields table, I don't get a confirmation dialog:

![jan-31-2017 03-09-43](https://cloud.githubusercontent.com/assets/51061/22462619/e30c937c-e762-11e6-8db7-563b4da137ba.gif)

Contrast this with performing a similar delete action on a scripted field:

![jan-31-2017 03-10-39](https://cloud.githubusercontent.com/assets/51061/22462633/f0601c24-e762-11e6-8075-8312ad545fee.gif)

For a consistent UX, we should have the delete in meta fields pop up a confirmation dialog just like the delete in scripted fields.

hickeyma commented on src/fixtures/logstash_fields.js:
```diff
@@ -30,6 +30,10 @@ function stubbedLogstashFields() {
     ['script number',     'number',     false, false, true,  false, { script: '1234' } ],
     ['script date',       'date',       false, false, true,  false, { script: '1234', lang: 'painless' } ],
     ['script murmur3',    'murmur3',    false, false, true,  false, { script: '1234' } ],
+    ['meta string',       'string',     false, false, true,  false, { meta: true, script: '\'i am a string\'', lang: 'expression' }],
```
Sure, removed

hickeyma commented on src/ui/public/field_editor/field_editor.html:
```diff
@@ -11,7 +11,7 @@
       ng-model="editor.field.name"
       required
-      placeholder="New Scripted Field"
+      placeholder="New Field"
```
I agree. Updated.

hickeyma commented on src/ui/public/index_patterns/_index_pattern.js:
```diff
@@ -44,6 +46,7 @@ export default function IndexPatternFactory(Private, Notifier, config, kbnIndex,
     intervalName: 'keyword',
     fields: 'json',
     sourceFilters: 'json',
+    metaFieldsVersion: 'boolean',
```
Ok, done.

hickeyma commented:
@ycombinator thanks for https://github.com/elastic/kibana/pull/8554#pullrequestreview-19285472 and https://github.com/elastic/kibana/pull/8554#issuecomment-276336528 . Update and read for review again.

ycombinator commented:
@spalger This PR LGTM; you want to take another look and merge if all looks good to you?

spalger commented on src/ui/public/index_patterns/_index_pattern.js:
```diff
@@ -110,7 +113,11 @@ export default function IndexPatternFactory(Private, Notifier, config, kbnIndex,
 
     if (!indexPattern.fields || !containsFieldCapabilities(indexPattern.fields)) {
       promise = indexPattern.refreshFields();
+    } else if (indexPattern.metaFieldsVersion <= 0) {
```
Changing this to `<=` prevents the check from catching undefined metaFieldsVersion. Reverting it to a falsy `!indexPattern.metaFieldsVersion` check should fix it.

hickeyma replied:
Ok, changed to check for false. Unfortunately had to change the type to 'boolean' as having issue with 'number'. Getting this error when index is first created: [mapper_parsing_exception] No handler for type [number] declared on field [metaFieldsVersion]

Any idea what is the issue is?

ycombinator commented:
@spalger wrote:

> One side-quest: refreshing fields currently restores all of the default metafields, even if you have already deleted them. Do you think it's feasible to prevent that from happening? If we are going to use "automatic field refreshing" as a way to upgrade index patterns when new changes are made (like we are doing here) then the refreshes should be non-destructive (a double negative in this case, but you get what I mean).

I thought about this some more and it got me wondering: why do we allow meta fields to be manually created and therefore deleted as well? Aren't they always going to be a subset of the standard fields (plus any meta fields defined by the `metaFields` advanced setting)? If so, then I think they should follow similar semantics as the standard fields. 

Moreover, if I understand the intent of this PR and the issue it resolves, it is about changing the _type_ of meta fields, not necessarily about creating new meta fields.

So I guess what I'm proposing is this: we should only allow one way for new meta fields to be created: via the `metaFields` advanced setting. Then, whenever a user refreshes (or creates anew) an index pattern, whatever meta fields are supposed to be there are shown in the meta fields tab introduced in this PR. Users may edit them but never delete them.

> Finally, the inclusion of the "aggregatable" and "searchable" checkboxes was requested because Kibana doesn't know about the metafield, or whether it can/should be used in searches or aggregations. The logic right now makes all meta fields searchable and aggregatable, but I think we need to restore the checkboxes put it on the user to decide. Of course, the checkboxes will need to actually persist if they are brought back.

I thought Kibana could tell whether a meta field is searchable or aggregatable by querying ES, per [your last response in this comment](https://github.com/elastic/kibana/pull/8554#issuecomment-260527560).

spalger commented:
> why do we allow meta fields to be manually created and therefore deleted as well? Aren't they always going to be a subset of the standard fields (plus any meta fields defined by the metaFields advanced setting)? If so, then I think they should follow similar semantics as the standard fields. ... we should only allow one way for new meta fields to be created: via the metaFields advanced setting.

I've been progressing with this change because it allows users to use metaFields that Kibana doesn't know about and allows having different meta fields per index.

> I thought Kibana could tell whether a meta field is searchable or aggregatable by querying ES, per your last response in this comment.

huh, I guess that is the case!

ycombinator commented:
Chatted with @spalger about this off PR to expedite the discussion a bit around the points I raised. The summary is that I agree it's worth providing users with the capability to define meta fields _per index_, not just "globally" via Advanced Settings.

Given that, I agree with Spencer's suggestion that refreshing an index pattern should *not* refresh the meta fields defined on the index pattern. In other words, an index pattern gets its _initial_ list of meta fields from the list in Advanced Settings _exactly once, upon index pattern creation_. To that end, we should also add that clarification to the help text for `metaFields` in Advanced Settings.

hickeyma commented:
@spalger, @ycombinator Comments updated and ready again for review.

ycombinator commented:
I know `metaFieldsVersion: 'number'` was changed to `metaFieldsVersion: 'boolean'` because there was some kind of a mapping? error (I can't seem to find its mention anywhere in the PR conversation). I think we need to get that to be a `number` again otherwise it would be difficult to change it later. @hickeyma could you paste that error again here, please?

hickeyma commented:
@ycombinator I agree that it should be 'number' but not sure this issue when change it. This is the error if it is number and when creating index for first time: [mapper_parsing_exception] No handler for type [number] declared on field [metaFieldsVersion]

ycombinator commented:
@hickeyma, I just got back to testing this PR and am noticing two things:

1. I don't see the `metaFieldsVersion` anywhere in the `.kibana` mapping at http://localhost:9200/.kibana/index-pattern/_mapping. I even tried deleting the index and letting Kibana recreate it but that property didn't show up.

2. I'm noticing this bug when refreshing the index patterns page after creating my first index pattern:

![feb-27-2017 22-33-48](https://cloud.githubusercontent.com/assets/51061/23394293/2b221674-fd3d-11e6-9b03-d893e4e2ab73.gif)

hickeyma commented:
@ycombinator thanks for https://github.com/elastic/kibana/pull/8554#issuecomment-282955691 . Updated and ready for review again.

ycombinator commented:
@spalger I'm good with this PR now. Mind re-reviewing as well? Thanks!

hickeyma commented:
@spalger, lets bring it home! :)

hickeyma commented:
@spalger, thanks for feedback in https://github.com/elastic/kibana/pull/8554#pullrequestreview-28471662.

Comments as follows:
>eslint's no-unused rule just got turned on so this has some new lint errors

Updated in https://github.com/elastic/kibana/pull/8554/commits/368da4126507e8746cecc84b1ad217ea7b1f2ed7

>why is _source not treated like a meta field?

If you make it a meta field only then there are issues because there seems to be a lot of references to '_source' field in the code. A comment in https://github.com/elastic/kibana/blob/master/src/ui/settings/defaults.js#L67 seems to imply it is more than a meta field. Maybe one of the Elastic folks might be able to shed some light on this? 

>when an index pattern is upgraded from to have special meta fields the previous fields are duplicated

What are the steps to reproduce this?

spalger commented:
> there seems to be a lot of references to '_source' field in the code

This is a good point, it's not really optional

>> when an index pattern is upgraded from to have special meta fields the previous fields are duplicated
> What are the steps to reproduce this?

Start kibana based on master, create an index pattern, checkout this branch, reload kibana. The index pattern will now have `_index` and friends in both the fields list and meta fields list:

![image](https://cloud.githubusercontent.com/assets/1329312/25051339/f0a9e53e-2100-11e7-91db-2b213b705ffd.png)

hickeyma commented:
@spalger, fix in for:
>when an index pattern is upgraded from to have special meta fields the previous fields are duplicated

Ready for review again.

hickeyma commented:
@spalger, some updates required after rebase with master.

hickeyma commented:
@spalger are we nearly getting there?

tsullivan commented:
Hi @hickeyma, I'm not sure why this went quiet, and I apologize. This PR now has a few conflicts that need to be resolved

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/127/)

nreese commented:
closing do to inactivity
